main.go

```diff
@@ -721,32 +740,34 @@ func (c *controller) saveHashring(ctx context.Context, hashring []receive.Hashri
 	return nil
 }
 
-func (c *controller) annotatePods(ctx context.Context, hashrings []receive.HashringConfig, statefulsets map[string]*appsv1.StatefulSet) {
-	for _, h := range hashrings {
-		if sts, exists := statefulsets[h.Hashring]; exists {
-			for i := 0; i < int(*sts.Spec.Replicas); i++ {
-				podName := fmt.Sprintf("%s-%d", sts.Name, i)
+func (c *controller) annotatePods(ctx context.Context) {
+	annotationKey := fmt.Sprintf("%s/%s", c.options.labelKey, "lastControllerUpdate")
 
-				pod, err := c.klient.CoreV1().Pods(c.options.namespace).Get(ctx, podName, metav1.GetOptions{})
-				if err != nil {
-					level.Error(c.logger).Log("msg", "failed to get pod", "err", err)
-				}
-
-				annotations := pod.Annotations
-				if annotations == nil {
-					annotations = make(map[string]string)
-				}
+	// Select pods that have a controllerLabel matching ours.
+	podList, err := c.klient.CoreV1().Pods(c.options.namespace).List(ctx,
+		metav1.ListOptions{
+			LabelSelector: fmt.Sprintf("%s=%s", c.options.labelKey, c.options.labelValue),
+		})
+	if err != nil {
+		level.Error(c.logger).Log("msg", "failed to list pods belonging to controller", "err", err)
+	} else {
+		for _, pod := range podList.Items {
+			podObj := pod.DeepCopy()
+
+			annotations := podObj.ObjectMeta.Annotations
+			if annotations == nil {
+				annotations = make(map[string]string)
+			}
 
-				annotations["annotationTimestamp"] = fmt.Sprintf("%d", time.Now().Unix())
-				pod.SetAnnotations(annotations)
+			annotations[annotationKey] = fmt.Sprintf("%d", time.Now().Unix())
```

Comment:
instead of iterating over the Pods it might make more sense to calculate the timestamp in advance and set all Pods to the same value as when the hashring changed?

Reply:
addressed with 0861153